WashU-IT-RIS/coldfront-wustl-fork
# ITDEV-34911; create default RIS-UserSupport group

This PR creates a new django app and associated coldfront command with the purpose of adding default groups and their associated permissions. Currently only the group `RIS-UserSupport` is defined for creation. Default permissions were obtained by adding all `project` and `allocation` permissions in Coldfront-QA. Those `id` and `codename` values were used to create two dictionary constants within the command's python file.

To add additional groups you must update the `RIS_DEFAULT_USER_GROUPS` dictionary in `coldfront.core.group.management.command.create_default_user_groups`. Default permissions can also be set in this dictionary.

Default permissions for `RIS-UserSupport` are stored in the `RIS_DEFAULT_USERSUPPORT_PERMISSIONS` dictionary. This is also stored in `coldfront.core.group.management.command.create_default_user_groups`.

A simple UnitTest was also written to verify that any groups defined in `RIS_DEFAULT_USER_GROUPS` are created with the expected permissions.

**TESTING INSTRUCTIONS**
**PART 1 - GROUP CREATION**
Case One:
1. Redeploy Coldfront cleanly via `docker-compose` to a local test machine.
   ```bash
   gh pr checkout 16
   docker compose build --no-cache && docker compose up -d
   ```
2. Perform Coldfront `initial_setup` as detailed in the README.md. This should trigger initial group creation. Also create a superuser to login with.
    ```bash
    docker exec -it coldfront-coldfront-workers-1 coldfront initial_setup
    docker exec -it coldfront-coldfront-workers-1 coldfront createsuperuser
    ```
3. Login to the Coldfront Admin site and verify the group now exists with expected permissions.
   ```bash
   firefox http://localhost:8080/admin
   ```
4. Navigate to the `Django Q/Scheduled tasks` page in the admin site. Confirm that the scheduled task `Grant RIS-UserSupport Manager status over all Projects` exists and is running successfully.
![image](https://github.com/user-attachments/assets/e998c39f-e1d6-4033-915e-cdcf613000a7)


Case Two:
1. Redeploy Coldfront wherever you have it running.
2. Exec into the container and manually trigger group creation.
   1. Docker Compose
      ```bash
      docker exec -it coldfront-coldfront-workers-1 coldfront create_default_user_groups
      ```
   2. GKE
      ```bash
      kubectl exec -it -n coldfront deploy/coldfront-deployment -- coldfront create_default_user_groups
      ```
3. Login to the Coldfront Admin site and verify the group now exists with the expected permissions.

Case Three:
1. Manually execute the UnitTests
   ```bash
   python -m venv coldfront-venv
   pip install -r requirements-dev.txt
   source coldfront-venv/bin/activate
   python manage.py test coldfront.core.group.tests
   ```


**PART 2 - SCHEDULED TASK EXECUTION**
Case One: 
1. Redeploy Coldfront cleanly via `docker-compose` to a local test machine.
   ```bash
   gh pr checkout 16
   docker compose build --no-cache && docker compose up -d
   ```
3. Obtain a backup of Coldfront-QA for testing
   ```bash
   # create backup in k8s pod
   kubectl exec -it -n coldfront statefulsets/coldfront-database-mariadb -- bash -c "mysqldump -u coldfront --password=nomoresecret coldfront | gzip > /bitnami/mariadb/data/dump-coldfrontqa.sql.gz"

   # copy backup to my Mac
   kubectl cp -n coldfront coldfront-database-mariadb-0:/bitnami/mariadb/data/dump-coldfrontqa.sql.gz $HOME/Documents/

   # copy backup into docker container
   docker cp $HOME/Documents/dump-coldfrontqa.sql.gz coldfront-coldfront-database-1:/tmp/dump-coldfrontqa.sql.gz

   # restore backup into MariaDB
   docker exec -it coldfront-coldfront-database-1 bash -c "gunzip < /tmp/dump-coldfrontqa.sql.gz | mariadb -u coldfront --password=nomoresecret coldfront"
   ```
4. Initialize local Coldfront DB with this backup.
   ```bash
   docker cp $HOME/Documents/dump-coldfrontqa.sql.gz coldfront-coldfront-database-1:/tmp/dump-coldfrontqa.sql.gz 
   docker exec -it coldfront-coldfront-database-1 bash -c "gunzip < /tmp/dump-coldfrontqa.sql.gz | mariadb -u coldfront --password=nomoresecret coldfront
   ```
5. Add the "new" scheduled tasks and groups for testing.
   ```bash
   docker exec -it coldfront-coldfront-workers-1 coldfront create_default_user_groups
   docker exec -it coldfront-coldfront-workers-1 coldfront add_scheduled_tasks
   ```
6. Login to the Coldfront Admin interface as the admin user.
   ```bash
   pass -c coldfront/qa/webui-admin
   firefox http://localhost:8080/admin
   ```
7. Identify any user(s) in the admin interface that you wish to use for testing. (I used the Gunnar and Tz-Kai users)
   - Add these users to the `RIS-UserSupport` user group. (Done on the user page)
   - Assign at least one of these users a password if they do not have one already; OIDC users will not have one by default. (I used `welcome123!`)
   ![image](https://github.com/user-attachments/assets/99eb7441-ba50-4ce1-ae53-b43228a90420)
8. Await scheduled execution of the `coldfront.core.group.tasks.grant_usersupport_global_project_manager` task or execute it manually.  
   ![image](https://github.com/user-attachments/assets/5b934beb-8777-4c4f-b276-5b9c7e7bedef)
9. Confirm that the user(s) now show up as Managers over all projects
   ![image](https://github.com/user-attachments/assets/d7ffb27d-7701-45cb-9447-a3717c53705a)
   ![image](https://github.com/user-attachments/assets/551d9832-2ee2-494d-b9f4-62b9e59d40ce)
